src/shared/network-cloud-util.c

```diff
@@ -0,0 +1,280 @@
+#include "parse-util.h"
+
+int network_cloud_manager_new(CloudProvider provider, const char *url, NetworkCloudManager **ret) {
+        _cleanup_(network_cloud_manager_freep) NetworkCloudManager *m = NULL;
+
+        m = new(NetworkCloudManager, 1);
+        if (!m)
+                return -ENOMEM;
+
+        *m = (NetworkCloudManager) {
+                .provider = provider,
+                .curl = curl_easy_init(),
+        };
+
+        if (!m->curl)
+                return -ENOSR;
+
+        if (url)
+                curl_easy_setopt(m->curl, CURLOPT_URL, url);
+
+        m->interfaces_by_mac = ordered_hashmap_new(&ether_addr_hash_ops);
+        if (!m->interfaces_by_mac)
+                return -ENOMEM;
+
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+void *network_cloud_manager_free(NetworkCloudManager *m) {
+        NetworkCloudMetaData *i;
+
+        if (!m)
+                return NULL;
+
+        if (m->curl)
+                curl_easy_cleanup(m->curl);
+
+        if (m->request_header)
+                curl_slist_free_all(m->request_header);
+
+
+        ORDERED_HASHMAP_FOREACH(i, m->interfaces_by_mac) {
+                ordered_hashmap_free_free(i->ipv4);
+                network_cloud_metadata_free(i);
+        }
+
+        ordered_hashmap_free(m->interfaces_by_mac);
+
+        return mfree(m);
+}
+
+int network_cloud_metadata_new(NetworkCloudMetaData **ret) {
+        _cleanup_(network_cloud_metadata_freep) NetworkCloudMetaData *m = NULL;
+
+        m = new0(NetworkCloudMetaData, 1);
+        if (!m)
+                return -ENOMEM;
+
+        m->ipv4 = ordered_hashmap_new(&trivial_hash_ops);
+        if (!m->ipv4)
+                return -ENOMEM;
+
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+void *network_cloud_metadata_free(NetworkCloudMetaData *m) {
+        if (!m)
+                return NULL;
+
+        return mfree(m);
+}
+
+static int format_cloud_to_ether(const char *s, char **ea) {
+        const char *m;
+        char *p, *l;
+        size_t i;
+
+        assert(s);
+
+        p = malloc0(strlen(s) + 128);
+        if (!p)
+                return -ENOMEM;
+
+        for (i = 0, l = p, m = s; i < strlen(s); i++) {
+                strncpy(l, m, 2);
+                l += 2;
+
+                if (m != s + strlen(s) - 2)
+                        *l++ = ':';
+                m += 2;
+        }
+
+        *ea = TAKE_PTR(p);
+        return 0;
+}
+
+size_t network_cloud_meta_data_write_callback(void *contents, size_t size, size_t nmemb, void *userdata) {
+        NetworkCloudManager *m = userdata;
+        size_t sz = size * nmemb;
+        int r = 0;
+
+        assert(contents);
+        assert(m);
+
+        if (!GREEDY_REALLOC(m->payload, m->payload_allocated, m->payload_size + sz)) {
+                r = log_oom();
+                goto fail;
+        }
+
+        memcpy(m->payload + m->payload_size, contents, sz);
+        m->payload_size += sz;
+
+        return sz;
+
+ fail:
```

Comment:
That `label` is unecessary, since there's only one jump point. But more significantly, negative `r` is cast to positive here, which doesn't look right.